src/os/bluestore/BlueStore.cc

```diff
@@ -8702,8 +8737,13 @@ void BlueStore::_kv_sync_thread()
       if (kv_stop)
 	break;
       dout(20) << __func__ << " sleep" << dendl;
-      kv_cond.wait(l);
+      std::cv_status status = kv_cond.wait_for(l,
+        std::chrono::milliseconds(int64_t(cct->_conf->bluestore_bluefs_balance_interval * 1000)));
       dout(20) << __func__ << " wake" << dendl;
+      if (status == std::cv_status::timeout) {
```

Comment:
Perhaps you can note the timeout, set some flag and proceed with the main loop. Which has to additionally check/reset the new flag in the above 'if' and go through 'committing' path calling regula _balance_bluefs_freespace. IMO this is safe.

Reply:
@ifed01: I will look at solution that sets flag and goes through processing part once more. I remember that I was unsure how it will behave when we do not have anything to process.